dhis2/dhis2-core
# feat: Push analytics job [DHIS2-16451]

### Summary
Adds a new job (type) `HTML_PUSH_ANALYTICS` to schedule sending the new HTML push analytics emails. I named it `HTML...` because there is already the _Push Analysis_ name in use for the old (text?) emial feature.

### Job Validation
With the new DB based scheduling I forgot to call the `Job#validate` method as part of running jobs. 
But having validation be a step prior to `Job#execute` now seemed inconsistent and unnecessary complicated in terms of getting a uniform error result and handling. Since the only jobs which actually had a `validate` phase were the ones checking the email configuration I updated these to make this check first thing in their `execute` method.

### Automatic Testing
Added a integration tests that makes sure the job parameters are stored and reconstructed correctly.

### Manual Testing
Create a new push analytics job, `POST` to `/api/41/jobConfigurations`:
```json
{
  "name": "test-push-analytics",
  "jobType": "HTML_PUSH_ANALYTICS",
  "jobParameters": {
    "dashboard": "nghVC4wtyzi",
    "receivers": "wl5cDMuUhmF",
    "mode":"RECEIVER"
  },
  "cronExpression": "0 0 9 ? * *"
}
```
Note that the `dashboard` UID should be an existing dashboard and the `receivers` an existing user group UID.

Setup the setting for the URL used for the push analytics service, for testing purposes I used `POST` to `/api/systemSettings/keyHtmlPushAnalyticsUrl` with body `http://localhost:8080/api/ping?username={username}&dashboard={id}`. The important part is that it is a URL with the `?username={username}&dashboard={id}`. I used `ping` as it lets you add the parameters without resulting in an error when called. 

Provide email settings in the system settings app and execute the job manually in the scheduler app. 
Check the `GET` `/api/jobConfigurations/<UID>/progress` for the outcome.
